File: kernels/Rust/compositron/src/dbs/dbspectrum.rs
// NOTE:
// - spectrum counts represent the center of their bins
// - integration happens in channel space, not in energy space

use std::collections::HashMap;

use thiserror::Error;

use crate::constants::M_E_KEV;
use crate::core::utils::{
    spectrum_match, EnergyDetector, LinearCalibration, LossyIntoF64, Spectrum,
    EcalCorrectionOrder, Unit
};
use crate::core::fitting::{self, VarproFitError};
use crate::dbs::fitting::*;

#[derive(Debug, Error)]
pub enum AnalysisError {
    #[error("Input parameter violates constraint {violated_constraint}")]
    InputError {
        violated_constraint: String,
    },

    #[error(
        "Attempting to integrate an area larger than the spectrum or \
        extracted peak"
    )]
    OutOfBounds,

    #[error(
        "Attempting analysis on background subtracted peak, but no subtraction \
        has been performed yet"
    )]
    NoPeakExtracted,

    #[error("Lineshape numerator or denominator areas overlap")]
    OverlappingAreas,

    #[error("Error during fitting")]
    FitError {
        #[from]
        inner: VarproFitError,
    },

    #[error("Could not convert units of eres to keV due to missing eres")]
    MissingEnergyResolution,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum PeakModel {
    Gauss,
    ErfLinear1Gauss,
    ErfLinear2Gauss,
    ErfLinear3Gauss,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum BinIntegrationScheme {
    Const,
    Linear,
}

#[derive(Debug, Clone, Copy)]
pub enum Area {
    Peak {
        width: Unit,
    },
    PeakWithOffset {
        width: Unit,
        offset: Unit,
    },
    Spectrum {
        left_bnd: f64,
        right_bnd: f64,
    },
}

impl Area {
    pub fn to_bnds_kev(&self, eres: Option<f64>) -> Option<(f64, f64)> {
        match self {
            Area::Peak { width } => {
                let width_kev = width.to_kev(eres)?;
                Some((M_E_KEV - width_kev / 2., M_E_KEV + width_kev / 2.))
            },
            Area::PeakWithOffset { width, offset } => {
                let width_kev = width.to_kev(eres)?;
                let offset_kev = offset.to_kev(eres)?;
                Some((
                    M_E_KEV - width_kev / 2. + offset_kev,
                    M_E_KEV + width_kev / 2. + offset_kev
                ))
            },
            Area::Spectrum { left_bnd, right_bnd } => {
                Some((*left_bnd, *right_bnd))
            },
        }
    }
}

#[derive(Debug, Clone)]
pub struct LineshapeParam {
    pub val: f64,
    pub err: f64,
    pub num: Vec<Area>,
    pub denom: Vec<Area>,
    pub in_corrected_peak: bool,
}

#[derive(Debug)]
pub struct LineshapeParamDefinition<'a> {
    pub name: &'a str,
    pub num: &'a [Area],
    pub denom: &'a [Area],
    pub in_corrected_peak: bool,
}

pub const STD_LINESHAPE_PARAMS: &[LineshapeParamDefinition; 4] = &[
    LineshapeParamDefinition {
        name: "S",
        num: &[Area::Peak { width: Unit::KeV(1.) }],
        denom: &[Area::Peak { width: Unit::KeV(20.) }],
        in_corrected_peak: true,
    },
    LineshapeParamDefinition {
        name: "W",
        num: &[
            Area::PeakWithOffset { width: Unit::KeV(1.), offset: Unit::KeV(3.) },
            Area::PeakWithOffset { width: Unit::KeV(1.), offset: Unit::KeV(-3.) },
        ],
        denom: &[Area::Peak { width: Unit::KeV(20.) }],
        in_corrected_peak: true,
    },
    LineshapeParamDefinition {
        name: "V/P",
        num: &[Area::Spectrum { left_bnd: 400., right_bnd: 500. }],
        denom: &[Area::Spectrum { left_bnd: 501., right_bnd: 521. }],
        in_corrected_peak: false,
    },
    LineshapeParamDefinition {
        name: "P/T",
        num: &[Area::Spectrum { left_bnd: 501., right_bnd: 521. }],
        denom: &[Area::Spectrum {
            left_bnd: f64::NEG_INFINITY, right_bnd: f64::INFINITY
        }],
        in_corrected_peak: false,
    },
];

pub struct DBSpectrum {
    pub spectrum: Spectrum,
    pub detector: EnergyDetector,
    pub counts: u64,
    pub dcounts: f64,
    pub peak: Option<Vec<f64>>,
    pub peak_bnds: Option<(usize, usize)>,
    pub peak_counts: f64,
    pub dpeak_counts: f64,
    pub peak_params: HashMap<&'static str, fitting::SimpleFitParam>,
    pub lineshape_params: HashMap<String, LineshapeParam>,
}

impl DBSpectrum {
    pub fn new(
        spectrum: Spectrum,
        detector: EnergyDetector,
    ) -> Self {
        let counts: u64 = spectrum_match!(
            &spectrum, arr => arr.iter().map(|&x| x as u64).sum()
        );

        let dcounts = (counts as f64).sqrt();

        DBSpectrum {
            spectrum,
            detector,
            counts,
            dcounts,
            peak_counts: f64::NAN,
            dpeak_counts: f64::NAN,
            peak: None,
            peak_bnds: None,
            peak_params: HashMap::new(),
            lineshape_params: HashMap::new(),
        }
    }

    pub fn default_analyze(&mut self) -> Result<(), AnalysisError> {
        self.correct_ecal(EcalCorrectionOrder::First)?;

        self.extract_peak(30., true, PeakModel::ErfLinear1Gauss)?;

        for param in STD_LINESHAPE_PARAMS {
            self.calc_lineshape_param(param, BinIntegrationScheme::Const)?;
        }

        self.peak = None;
        self.peak_bnds = None;

        Ok(())
    }

    fn get_peak_energies(&self) -> Vec<f64> {
        let ecal = self.detector.corrected_ecal.unwrap_or(self.detector.ecal);

        let bnds = self.peak_bnds.unwrap();

        (bnds.0..=bnds.1).map(|i| ecal.from_index(i)).collect::<Vec<f64>>()
    }

    fn fit_peak(
        &mut self, peak_model: PeakModel,
    ) -> Result<(), AnalysisError> {
        let x = self.get_peak_energies();
        let y = self.peak.as_ref().unwrap();

        match peak_model {
            PeakModel::Gauss => {
                self.peak_params = fit_gaussian(&x, &y, vec![511., 1.])?;
            },
            PeakModel::ErfLinear1Gauss => {
                self.peak_params = fit_erf_linear_1_gauss(&x, &y)?;
            },
            PeakModel::ErfLinear2Gauss => {
                self.peak_params = fit_erf_linear_2_gauss(&x, &y)?;
            },
            PeakModel::ErfLinear3Gauss => {
                self.peak_params = fit_erf_linear_3_gauss(&x, &y)?;
            },
        }

        Ok(())
    }

    fn subtract_bg(
        &mut self, peak_model: PeakModel
    ) -> Result<(), AnalysisError> {
        self.fit_peak(peak_model)?;

        let x = self.get_peak_energies();
        let y = self.peak.as_mut().unwrap();

        let corr = background(
            &x,
            self.peak_params.get("erf_amp").unwrap().val,
            self.peak_params.get("x0_1").unwrap().val,
            self.peak_params.get("sig_1").unwrap().val,
            self.peak_params.get("lin").unwrap().val,
            self.peak_params.get("const").unwrap().val,
        );

        for (s, bg) in y.iter_mut().zip(corr) {
            *s = *s - bg;
        }

        Ok(())
    }

    pub fn extract_peak(
        &mut self, peak_width: f64, bg_corr: bool, peak_model: PeakModel,
    ) -> Result<&[f64], AnalysisError> {
        let ecal = self.detector.corrected_ecal.unwrap_or(self.detector.ecal);

        let left_peak_idx = ecal.to_index_rounded(M_E_KEV - peak_width / 2.);

        let right_peak_idx = ecal.to_index_rounded(M_E_KEV + peak_width / 2.);

        self.peak_bnds = Some((left_peak_idx, right_peak_idx));

        self.peak = Some(spectrum_match!(
            &self.spectrum,
            arr => arr.rows(left_peak_idx, right_peak_idx - left_peak_idx + 1)
                .iter()
                .map(|&x| x as f64)
                .collect()
        ));

        if bg_corr == true {
            self.subtract_bg(peak_model)?;
        }

        Ok(self.peak.as_ref().unwrap())
    }

    pub fn correct_ecal(
        &mut self, order: EcalCorrectionOrder
    ) -> Result<LinearCalibration, AnalysisError> {
        if order == EcalCorrectionOrder::None {
            return Ok(self.detector.ecal);
        }

        if self.peak == None {
            self.extract_peak(20., false, PeakModel::Gauss)?;
        }

        self.fit_peak(PeakModel::Gauss)?;

        let c = self.peak_params.get("x0_1").unwrap().val;

        let mut ecal = self.detector.ecal;

        match order {
            EcalCorrectionOrder::Zeroth => {
                ecal.offset += M_E_KEV - c;
            },
            EcalCorrectionOrder::First => {
                ecal.scale *= (M_E_KEV - ecal.offset) / (c - ecal.offset);
            },
            EcalCorrectionOrder::None => unreachable!(),
        }

        self.detector.corrected_ecal = Some(ecal);

        Ok(ecal)
    }

    fn integrate_generic<T: LossyIntoF64>(
        spectrum: &[T],
        mut lower_ch_bnd: f64,
        mut upper_ch_bnd: f64,
        bin_integration_scheme: BinIntegrationScheme,
    ) -> Result<f64, AnalysisError> {
        // TODO: We could allow for a lower_ch_bdn of >= -0.5
        // but I'm too lazy to deal with this special edge case now
        if lower_ch_bnd.is_infinite() {
            lower_ch_bnd = 0.;
        }
        if upper_ch_bnd.is_infinite() {
            upper_ch_bnd = spectrum.len() as f64 - 1.;
        }

        if lower_ch_bnd < 0. || lower_ch_bnd > spectrum.len() as f64 - 1. {
            return Err(AnalysisError::OutOfBounds);
        }
        if upper_ch_bnd < 0. || upper_ch_bnd > spectrum.len() as f64 - 1. {
            return Err(AnalysisError::OutOfBounds);
        }

        let mut counts: f64 = 0.;

        match bin_integration_scheme {
            BinIntegrationScheme::Const => {
                // index of lowest channel still inside ROI
                let lowest_ch = lower_ch_bnd.round() as usize;
                // index of highest channel still inside ROI
                let highest_ch = upper_ch_bnd.round() as usize;

                counts += spectrum[lowest_ch..=highest_ch]
                    .iter()
                    .map(|&x| T::lossy_into_f64(x))
                    .sum::<f64>();

                // Left edge counts

                let lowest_ch_counts = T::lossy_into_f64(spectrum[lowest_ch]);
                let x1 = lowest_ch as f64 - 0.5;
                counts -= lowest_ch_counts * (lower_ch_bnd - x1);

                // Right edge counts

                let highest_ch_counts = T::lossy_into_f64(spectrum[highest_ch]);
                let x2 = highest_ch as f64 + 0.5;
                counts -= highest_ch_counts * (x2 - upper_ch_bnd);
            },
            BinIntegrationScheme::Linear => {
                // index of lowest channel still inside ROI
                let lowest_ch = lower_ch_bnd.floor() as usize;
                // index of highest channel still inside ROI
                let highest_ch = upper_ch_bnd.ceil() as usize;

                counts += spectrum[lowest_ch+1..highest_ch]
                    .iter()
                    .map(|&x| T::lossy_into_f64(x))
                    .sum::<f64>();

                counts += 0.5 * T::lossy_into_f64(spectrum[lowest_ch]);
                counts += 0.5 * T::lossy_into_f64(spectrum[highest_ch]);

                // Left edge counts

                let x1 = lowest_ch as f64;

                let y1 = T::lossy_into_f64(spectrum[lowest_ch]);
                let y2 = T::lossy_into_f64(spectrum[lowest_ch + 1]);

                let y_l = (y2 - y1) * (lower_ch_bnd - x1) + y1;

                counts -= 0.5 * (y1 + y_l) * (lower_ch_bnd - x1);

                // Right edge counts

                let x1 = (highest_ch - 1) as f64;
                let x2 = highest_ch as f64;

                let y1 = T::lossy_into_f64(spectrum[highest_ch - 1]);
                let y2 = T::lossy_into_f64(spectrum[highest_ch]);

                let y_u = (y2 - y1) * (upper_ch_bnd - x1) + y1;

                counts -= 0.5 * (y_u + y2) * (x2 - upper_ch_bnd);
            },
        }

        Ok(counts)
    }

    pub fn integrate(
        &self,
        area: Area,
        in_corrected_peak: bool,
        bin_integration_scheme: BinIntegrationScheme,
    ) -> Result<f64, AnalysisError> {
        let (lower_bnd, upper_bnd) = area.to_bnds_kev(self.detector.eres)
            .ok_or(AnalysisError::MissingEnergyResolution)?;

        if in_corrected_peak && self.peak == None {
            return Err(AnalysisError::NoPeakExtracted);
        }
        if lower_bnd >= upper_bnd {
            return Err(AnalysisError::InputError {
                violated_constraint: format!(
                    "Lower bound < upper bound (got {lower_bnd} < {upper_bnd})"
                )
            });
        }

        let ecal = self.detector.corrected_ecal.unwrap_or(self.detector.ecal);

        let mut lower_ch_bnd = ecal.to_index_f64(lower_bnd);
        let mut upper_ch_bnd = ecal.to_index_f64(upper_bnd);

        if in_corrected_peak {
            let peak_bnds = self.peak_bnds.unwrap();
            lower_ch_bnd -= peak_bnds.0 as f64;
            upper_ch_bnd -= peak_bnds.0 as f64;
            let peak = self.peak.as_ref().unwrap();
            DBSpectrum::integrate_generic(
                peak, lower_ch_bnd, upper_ch_bnd, bin_integration_scheme,
            )
        } else {
            spectrum_match!(
                &self.spectrum, spectrum => DBSpectrum::integrate_generic(
                    spectrum.as_slice(),
                    lower_ch_bnd,
                    upper_ch_bnd,
                    bin_integration_scheme,
                )
            )
        }
    }

    fn get_area_bnds(&self, areas: &[Area]) -> Result<Vec<(f64, f64)>, AnalysisError> {
        let bnds = areas
            .iter()
            .map(|area| area.to_bnds_kev(self.detector.eres))
            .collect::<Option<Vec<_>>>()
            .ok_or(AnalysisError::MissingEnergyResolution)?;

        for i in 0..areas.len() - 1 {
            for j in i + 1..areas.len() {
                if !(bnds[i].1 < bnds[j].0 || bnds[j].1 < bnds[i].0) {
                    return Err(AnalysisError::OverlappingAreas);
                }
            }
        }

        Ok(bnds)
    }

    pub fn calc_lineshape_param(
        &mut self,
        definition: &LineshapeParamDefinition,
        bin_integration_scheme: BinIntegrationScheme,
    ) -> Result<&LineshapeParam, AnalysisError> {
        let num_bnds = self.get_area_bnds(definition.num)?;
        let denom_bnds = self.get_area_bnds(definition.denom)?;

        let num = definition.num
            .iter()
            .map(|&area| self.integrate(
                area, definition.in_corrected_peak, bin_integration_scheme
            ))
            .collect::<Result<Vec<_>, _>>()?
            .iter()
            .sum::<f64>();

        let denom = definition.denom
            .iter()
            .map(|&area| self.integrate(
                area, definition.in_corrected_peak, bin_integration_scheme
            ))
            .collect::<Result<Vec<_>, _>>()?
            .iter()
            .sum::<f64>();

        let mut common = 0.;

        for num_bnd in num_bnds.iter() {
            for denom_bnd in denom_bnds.iter() {
                let left_bnd = num_bnd.0.max(denom_bnd.0);
                let right_bnd = num_bnd.1.min(denom_bnd.1);
                if left_bnd < right_bnd {
                    common += self.integrate(
                        Area::Spectrum { left_bnd, right_bnd },
                        definition.in_corrected_peak,
                        bin_integration_scheme,
                    )?;
                }
            }
        }

        let val = num / denom;
        let err = val * (
            (num - common) / num.powi(2) +
            (denom - common) / denom.powi(2) +
            common * ((denom - num) / (denom * num)).powi(2)
        ).sqrt();

        let ls_param = LineshapeParam {
            val,
            err,
            num: definition.num.to_owned(),
            denom: definition.denom.to_owned(),
            in_corrected_peak: definition.in_corrected_peak,
        };

        self.lineshape_params.insert(definition.name.into(), ls_param);

        Ok(self.lineshape_params.get(definition.name).unwrap())
    }
}

File: kernels/Python/compositron/src/compositron/dbs/dbspectrum.py
from copy import copy, deepcopy
from dataclasses import dataclass
from enum import Enum, auto

import numpy as np
import numpy.typing as npt

from ..constants import M_E_KEV
from ..core.fitting import SimpleFitParam
from ..core.utils import (
    EcalCorrectionOrder, to_min_type_spectrum, Spectrum, EnergyDetector, Unit,
    KeV
)
from .fitting import (
    erf_linear_background, fit_gaussian, fit_erf_linear_1_gauss,
    fit_erf_linear_2_gauss, fit_erf_linear_3_gauss,
)


class PeakModel(Enum):
    GAUSS = auto()
    ERFLINEAR1GAUSS = auto()
    ERFLINEAR2GAUSS = auto()
    ERFLINEAR3GAUSS = auto()


class BinIntegrationScheme(Enum):
    CONST = auto()
    LINEAR = auto()


@dataclass
class PeakArea:
    width: Unit

    def to_bnds_kev(self, eres: None | float) -> tuple[float, float]:
        width_kev = self.width.to_kev(eres)
        return (M_E_KEV - width_kev / 2, M_E_KEV + width_kev / 2)


@dataclass
class PeakAreaWithOffset:
    width: Unit
    offset: Unit

    def to_bnds_kev(self, eres: None | float) -> tuple[float, float]:
        width_kev = self.width.to_kev(eres)
        offset_kev = self.offset.to_kev(eres)
        return (
            M_E_KEV - width_kev / 2 + offset_kev,
            M_E_KEV + width_kev / 2 + offset_kev
        )


@dataclass
class SpectrumArea:
    left_bnd: float
    right_bnd: float

    def to_bnds_kev(self, _: None | float) -> tuple[float, float]:
        return (self.left_bnd, self.right_bnd)


Area = PeakArea | PeakAreaWithOffset | SpectrumArea


@dataclass
class LineshapeParam:
    val: float
    err: float
    num: list[Area]
    denom: list[Area]
    in_corrected_peak: bool


@dataclass
class LineshapeParamDefinition:
    name: str
    num: list[Area]
    denom: list[Area]
    in_corrected_peak: bool


STD_LINESHAPE_PARAMS: list[LineshapeParamDefinition] = [
    LineshapeParamDefinition(
        name = "S",
        num = [PeakArea(KeV(1))],
        denom = [PeakArea(KeV(20))],
        in_corrected_peak = True,
    ),
    LineshapeParamDefinition(
        name = "W",
        num = [
            PeakAreaWithOffset(KeV(1), KeV(3)),
            PeakAreaWithOffset(KeV(1), KeV(-3))
        ],
        denom = [PeakArea(KeV(20))],
        in_corrected_peak = True,
    ),
    LineshapeParamDefinition(
        name = "V/P",
        num = [SpectrumArea(400, 500)],
        denom = [SpectrumArea(501, 521)],
        in_corrected_peak = False,
    ),
    LineshapeParamDefinition(
        name = "P/T",
        num = [SpectrumArea(501, 521)],
        denom = [SpectrumArea(-np.inf, np.inf)],
        in_corrected_peak = False,
    ),
]


@dataclass
class DBSpectrum:
    spectrum: Spectrum
    detector: EnergyDetector
    counts: int
    dcounts: float
    peak: None | npt.NDArray[np.floating]
    peak_bnds: None | tuple[int, int]
    peak_counts: None | float
    dpeak_counts: None | float
    peak_params: dict[str, SimpleFitParam]
    lineshape_params: dict[str, LineshapeParam]

    def __init__(
        self, spectrum: npt.ArrayLike, detector: EnergyDetector,
    ) -> None:
        counts = np.sum(spectrum)
        dcounts = np.sqrt(counts)

        self.spectrum = to_min_type_spectrum(spectrum)
        self.detector = detector
        self.counts = counts
        self.dcounts = dcounts
        self.peak = None
        self.peak_bnds = None
        self.peak_counts = None
        self.dpeak_counts = None
        self.peak_params = {}
        self.lineshape_params = {}

    def default_analyze(self) -> None:
        self.correct_ecal(EcalCorrectionOrder.FIRST)

        self.extract_peak(30, PeakModel.ERFLINEAR1GAUSS)

        for param in STD_LINESHAPE_PARAMS:
            self.calc_lineshape_param(param, BinIntegrationScheme.CONST)

        self.peak = None
        self.peak_bnds = None

    def _get_peak_energies(self) -> npt.NDArray[np.float64]:
        ecal = self.detector.corrected_ecal or self.detector.ecal

        assert (bnds := self.peak_bnds) is not None

        return (
            np.arange(bnds[0], bnds[1], dtype=np.float64) * ecal.scale
            + ecal.offset
        )

    def _fit_peak(self, peak_model: PeakModel) -> None:
        x = self._get_peak_energies()
        assert (y := self.peak) is not None

        m = np.max(y)

        match peak_model:
            case PeakModel.GAUSS:
                self.peak_params = fit_gaussian(x, y, [m, 511, 1])
            case PeakModel.ERFLINEAR1GAUSS:
                self.peak_params = fit_erf_linear_1_gauss(
                    x, y, [m, 511, 1, -10, 0, 0]
                )
            case PeakModel.ERFLINEAR2GAUSS:
                self.peak_params = fit_erf_linear_2_gauss(
                    x, y, [m / 2, 511, 1, m / 2, 511, 1.5, -10, 0, 0]
                )
            case PeakModel.ERFLINEAR3GAUSS:
                self.peak_params = fit_erf_linear_3_gauss(
                    x, y,
                    [m / 3, 511, 1, m / 3, 511, 1.5, m / 3, 511, 2, -10, 0, 0]
                )

    def _subtract_bg(self, peak_model: PeakModel) -> None:
        if peak_model == PeakModel.GAUSS:
            return

        self._fit_peak(peak_model)

        x = self._get_peak_energies()
        assert (y := self.peak) is not None

        corr = erf_linear_background(
            x,
            self.peak_params["erf_amp"].val,
            self.peak_params["x0_1"].val,
            self.peak_params["sig_1"].val,
            self.peak_params["lin"].val,
            self.peak_params["const"].val,
        )

        y -= corr

    def extract_peak(
        self, peak_width: float, peak_model: PeakModel
    ) -> npt.NDArray[np.float64]:
        ecal = self.detector.corrected_ecal or self.detector.ecal

        left_peak_idx = ecal.to_index_rounded(M_E_KEV - peak_width / 2)
        right_peak_idx = ecal.to_index_rounded(M_E_KEV + peak_width / 2)

        self.peak_bnds = (left_peak_idx, right_peak_idx)

        self.peak = (
            self.spectrum[left_peak_idx:right_peak_idx].astype(np.float64)
        )

        self._subtract_bg(peak_model)

        return self.peak

    def correct_ecal(self, order: EcalCorrectionOrder) -> None:
        if order == EcalCorrectionOrder.NONE:
            return

        if self.peak is None:
            self.extract_peak(20, PeakModel.GAUSS)

        self._fit_peak(PeakModel.GAUSS)

        c = self.peak_params["x0_1"].val

        ecal = copy(self.detector.ecal)

        match order:
            case EcalCorrectionOrder.ZEROTH:
                ecal.offset += M_E_KEV - c
            case EcalCorrectionOrder.FIRST:
                ecal.scale *= (M_E_KEV - ecal.offset) / (c - ecal.offset);

        self.detector.corrected_ecal = ecal

    @staticmethod
    def _integrate_const(
        spectrum: npt.NDArray, lower_ch_bnd: float, upper_ch_bnd: float
    ) -> float:
        counts = 0.

        lowest_ch = int(round(lower_ch_bnd))
        highest_ch = int(round(upper_ch_bnd))

        counts += np.sum(spectrum[lowest_ch:highest_ch+1], dtype=np.float64)

        # Left edge counts

        lowest_ch_counts = float(spectrum[lowest_ch])
        x1 = lowest_ch - 0.5
        counts -= lowest_ch_counts * (lower_ch_bnd - x1)

        # Right edge counts

        highest_ch_counts = float(spectrum[highest_ch])
        x2 = highest_ch + 0.5
        counts -= highest_ch_counts * (x2 - upper_ch_bnd)

        return counts

    @staticmethod
    def _integrate_linear(
        spectrum: npt.NDArray, lower_ch_bnd: float, upper_ch_bnd: float
    ) -> float:
        counts = 0.

        lowest_ch = int(np.floor(lower_ch_bnd))
        highest_ch = int(np.ceil(upper_ch_bnd))

        counts += np.sum(spectrum[lowest_ch+1:highest_ch], dtype=np.float64)

        counts += 0.5 * spectrum[lowest_ch]
        counts += 0.5 * spectrum[highest_ch]

        # Left edge counts

        x1 = float(lowest_ch)

        y1 = float(spectrum[lowest_ch])
        y2 = float(spectrum[lowest_ch + 1])

        y_l = (y2 - y1) * (lower_ch_bnd - x1) + y1

        counts -= 0.5 * (y1 + y_l) * (lower_ch_bnd - x1)

        # Left edge counts

        x1 = float(highest_ch - 1)
        x2 = float(highest_ch)

        y1 = float(spectrum[highest_ch - 1])
        y2 = float(spectrum[highest_ch])

        y_u = (y2 - y1) * (upper_ch_bnd - x1) + y1

        counts -= 0.5 * (y_u + y2) * (x2 - upper_ch_bnd)

        return counts

    @staticmethod
    def _integrate_generic(
        spectrum: npt.NDArray,
        lower_ch_bnd: float,
        upper_ch_bnd: float,
        bin_integration_scheme: BinIntegrationScheme,
    ) -> float:
        if np.isinf(lower_ch_bnd):
            lower_ch_bnd = 0.

        if np.isinf(upper_ch_bnd):
            upper_ch_bnd = float(len(spectrum) - 1)

        if (
            lower_ch_bnd < 0 or lower_ch_bnd > len(spectrum) - 1 or
            upper_ch_bnd < 0 or upper_ch_bnd > len(spectrum) - 1
        ):
            raise ValueError(
                "Attempting to integrate an area larger than the spectrum or "
                "extracted peak"
            )

        match bin_integration_scheme:
            case BinIntegrationScheme.CONST:
                return DBSpectrum._integrate_const(
                    spectrum, lower_ch_bnd, upper_ch_bnd
                )
            case BinIntegrationScheme.LINEAR:
                return DBSpectrum._integrate_linear(
                    spectrum, lower_ch_bnd, upper_ch_bnd
                )

    def integrate(
        self,
        area: Area,
        in_corrected_peak: bool,
        bin_integration_scheme: BinIntegrationScheme,
    ) -> float:
        lower_bnd, upper_bnd = area.to_bnds_kev(self.detector.eres)

        if in_corrected_peak and self.peak is None:
            raise RuntimeError(
                "Attempting analysis on background subtracted peak, but no "
                "subtraction has been performed yet"
            )

        if lower_bnd >= upper_bnd:
            raise ValueError(
                "Input parameter violates constraint "
                f"Lower bound < upper bound (got {lower_bnd} < {upper_bnd})"
            )

        ecal = self.detector.corrected_ecal or self.detector.ecal

        lower_ch_bnd = ecal.to_index_float(lower_bnd)
        upper_ch_bnd = ecal.to_index_float(upper_bnd)

        if in_corrected_peak:
            assert (peak_bnds := self.peak_bnds) is not None
            lower_peak_bnd, _ = peak_bnds

            lower_ch_bnd -= lower_peak_bnd
            upper_ch_bnd -= lower_peak_bnd

            assert (peak := self.peak) is not None

            return DBSpectrum._integrate_generic(
                peak, lower_ch_bnd, upper_ch_bnd, bin_integration_scheme,
            )
        else:
            return DBSpectrum._integrate_generic(
                self.spectrum,
                lower_ch_bnd,
                upper_ch_bnd,
                bin_integration_scheme,
            )

    def _get_area_bnds(self, areas: list[Area]) -> list[tuple[float, float]]:
        bnds = [area.to_bnds_kev(self.detector.eres) for area in areas]

        for i in range(len(areas) - 1):
            for j in range(i + 1, len(areas)):
                if not (bnds[i][1] < bnds[j][0] or bnds[j][1] < bnds[i][0]):
                    raise ValueError(
                        "Lineshape numerator or denominator areas overlap"
                    )

        return bnds

    def calc_lineshape_param(
        self,
        definition: LineshapeParamDefinition,
        bin_integration_scheme: BinIntegrationScheme,
    ) -> LineshapeParam:
        num_bnds = self._get_area_bnds(definition.num)
        denom_bnds = self._get_area_bnds(definition.denom)

        num = sum([
            self.integrate(
                area, definition.in_corrected_peak, bin_integration_scheme
            ) for area in definition.num
        ])
        denom = sum([
            self.integrate(
                area, definition.in_corrected_peak, bin_integration_scheme
            ) for area in definition.denom
        ])

        common = 0.

        for num_bnd in num_bnds:
            for denom_bnd in denom_bnds:
                left_bnd = max(num_bnd[0], denom_bnd[0])
                right_bnd = min(num_bnd[1], denom_bnd[1])

                if left_bnd < right_bnd:
                    common += self.integrate(
                        SpectrumArea(left_bnd, right_bnd),
                        definition.in_corrected_peak,
                        bin_integration_scheme,
                    )

        val = num / denom
        err = val * np.sqrt(
            (num - common) / num**2 +
            (denom - common) / denom**2 +
            common * ((denom - num) / (denom * num))**2
        )

        ls_param = LineshapeParam(
            val,
            err,
            deepcopy(definition.num),
            deepcopy(definition.denom),
            definition.in_corrected_peak,
        )

        self.lineshape_params[definition.name] = ls_param

        return ls_param
